### ISSUE TITLE
hybrid_cqt magnitude scaling is inconsistent

### ISSUE THREAD

**Issue by lostanlen** [bmcfee/librosa#341]
Hi,
With @CarmineCella we've tried running the CQT on synthetic sounds, such as white noise (`np.random.randn`) as well as Diracs.
It looks like the `cqt` function, which relies on the pyramid algorithm of Schoerkuber and Klapuri, is not appropriately scaled in magnitude across octaves.

Below is the code to compute the CQT of a Dirac and the corresponding figure:

```python
import librosa
import numpy as np
x = np.zeros(32768)
x[16384] = 1.0
sr = 22050.0
c = librosa.cqt(x, sr=sr)
librosa.display.specshow(c)
```

![cqt_dirac](https://cloud.githubusercontent.com/assets/3943142/14914832/178b1d5a-0e0e-11e6-98f2-2f87c4e8ff89.png)

Am I mistaken to think that the magnitudes should stay equal across octaves instead ?

**Comment by dpwe** [comment]
I think you can make an argument that the amplitude should drop for lower
frequencies, because the bandwidths are getting smaller, so the total
energy (from a constant-energy-per-Hz signal like an impulse) is smaller at
lower frequencies.  (You could argue the other way, too, but because real
audio tends to have ~ 1/f spectrum, the implicit pre-emphasis of
constant-peak-gain per band works quite well).

But the banding here is clearly an artifact.  I thought we dealt with this
before?

  DAn.

On Fri, Apr 29, 2016 at 7:33 AM, Vincent Lostanlen <notifications@github.com

**Comment by lostanlen** [comment]
> I think you can make an argument that the amplitude should drop for lower frequencies, because the bandwidths are getting smaller, so the total energy (from a constant-energy-per-Hz signal like an impulse) is smaller at lower frequencies.
OK, I see your point. Several renormalization conventions (1/f, 1/sqrt(f), and 1) coexist in the wavelet literature anyway.

But there seems to be an inconsistency within the CQT implementations in librosa: `pseudo_cqt` adopts the "1" convention, whereas `cqt` adopts the "1/f" convention. As a result, `hybrid_cqt` is difficult to interpret in terms of energy distribution.

As you said, the presence of sharp discontinuities at octave boundaries is also a problem, especially for percussive or environmental sounds, though not so much for pitched music or speech.

**Comment by dpwe** [comment]
I thin this is already fixed (see below).  But using an impulse nicely
illustrates the problems with letting hop size exceed the shrinking
time window for higher bands.

**Comment by bmcfee** [comment]
> I thin this is already fixed (see below).

I think I concur ; @lostanlen which version of librosa are you using?  The banding should have been fixed in 0.4.2.

I just tried your example code on my the current master branch and get very different results:
```python
plt.figure(figsize=(8, 6))
librosa.display.specshow(librosa.logamplitude(c**2, ref_power=np.max))
plt.colorbar()
plt.tight_layout()
```
![image](https://cloud.githubusercontent.com/assets/1190540/14917322/78d544a4-0dec-11e6-8dde-5cc4c8a9353d.png)

Note: this is using the new resampler, so YMMV if using scikits.samplerate or scipy resampling, but the differences should be small.

**Comment by lostanlen** [comment]
I was using librosa v0.4.1. After upgrading to librosa v0.4.2, I can reproduce @bmcfee's figure.
This was indeed solved in #307. Sorry for the inconvenience.

Mapping the amplitudes to a dB scale, like you did, makes the artifacts perfectly acceptable.
Still, there is a sharp 8x magnitude difference (18 dB) in `hybrid_cqt` at the frequential boundary between the two methods.

For the record, below is the magnitude (not in dB) at the impulse location with new librosa 0.4.2. Green is hybrid CQT, blue is CQT.

```python
c = librosa.cqt(x, sr=sr)
plt.plot(c[:, 32])
hybrid_c = librosa.hybrid_cqt(x, sr=sr)
plt.plot(hybrid_c[:, 32])
```
![hybrid_versus_cqt](https://cloud.githubusercontent.com/assets/3943142/14918164/b6a33364-0e22-11e6-8ed9-33c485e056a0.png)

**Comment by dpwe** [comment]
It's odd that it's not time-symmetric.  Setting x[0] to 1 looks about
right, x[-1], not so much:[image: Inline image 1]
I agree that the noise in the different HF bands is also a problem.

  DAn.

On Fri, Apr 29, 2016 at 9:57 AM, Vincent Lostanlen <notifications@github.com

**Comment by bmcfee** [comment]
@lostanlen I think I understand what's going on here.

There's probably two octaves worth of early downsampling hitting the CQT there, followed by one octave of pseudo-cqt, which adds up to a missing factor of 2**3=8 that you're seeing.  This seems like a bug, and should be easy to fix.  @ebattenberg have any thoughts on this?

The notches are a different issue, but they become less severe as the hop length decreases.  I'm not sure this can be resolved without using a redundant, multi-phase basis like K&S did.

**Comment by dpwe** [comment]
Yeah.

I moved the impulse all across the sample, and took the max of all the
CQTs.  The notches are consistent!  And ugly.

It does seem like an impulse should give a smooth response with
frequency.  It's actually better than I feared (I expected more
dependence of the amplitude on the exact time location), but it still
looks broken to me.

**Comment by bmcfee** [comment]
Okay, I see two issues in this thread then.

1. Hybrid cqt scaling is inconsistent. This is easy to fix.
2. High frequency cqt is lossy at high hop lengths. This is fixable by using redundant bases, which we probably should do in preparation for #165. However, it requires substantial changes to all cqt-related functions.

I think we can do 1 for the next minor release 0.4.3 and punt the redundant basis implementation to 0.5.

Sound good?

**Comment by dpwe** [comment]
Sounds good.

  DAn.

**Comment by ebattenberg** [comment]
All these scaling issues make my head hurt.  And I guess they're even harder to deal with in the hybrid_cqt because there are so many things that affect scale: FFT length, basis function normalization, how many times we do early downsampling, etc.  Then the combination of all those factors need to match up between pseudo_cqt and regular cqt.  

@bmcfee, I forget, have we been adding unit tests along the way as we address each of these issues?  

I agree that smooth scaling should be our goal.  I wonder if there's a convention we can agree to guide in choosing which arbitrary and consistent scale we use.  Can we choose something that has some sort of physical meaning for the original audio waveform?

**Comment by bmcfee** [comment]
> FFT length, basis function normalization, how many times we do early downsampling, etc. Then the combination of all those factors need to match up between pseudo_cqt and regular cqt.

Until I see otherwise, I'm going to blame everything on scale loss during downsampling ops.  The filter normalization stuff seems to be settled -- I experimented with this a bunch during the last round #307 , and I think within an individual FFT, everything works out smoothly.



> @bmcfee, I forget, have we been adding unit tests along the way as we address each of these issues?

Probably not as much as we should.  I'm keen to hear ideas on how to design good tests for this though.



> I agree that smooth scaling should be our goal. I wonder if there's a convention we can agree to guide in choosing which arbitrary and consistent scale we use. Can we choose something that has some sort of physical meaning for the original audio waveform?

Multiplying by `sqrt(ratio)` after a resample op makes physical sense if you want the signal to have equivalent energy before and after resampling.  CQT uses an additional factor of `sqrt(ratio)` to account for the fact that the basis has not been re-scaled; this effectively simulates having the correct filter normalization for each octave, and gives the correct (continuous) response when looking at CQT of white noise.

**Comment by bmcfee** [comment]
I did a bit of playing around, and I think we can fix hybrid scaling by multiplying the pseudo_cqt output by a factor of `2**(downsample_count - 1)` where `downsample_count` is as computed in `__early_downsample`.  This seems to give consistent responses (max over time) for `hop_length = 2**k, k in [6..10]`, eg for hop=256:

![image](https://cloud.githubusercontent.com/assets/1190540/14928077/b6fe18c8-0e23-11e6-9bc2-1a7a174a007e.png)


However, uneven hop multiples (eg, 320) yield a strange discontinuity:

![image](https://cloud.githubusercontent.com/assets/1190540/14928086/c97cf744-0e23-11e6-8e42-c0ad303c5187.png)

**Comment by lostanlen** [comment]
_[I wrote this message after the "sound good ?" of @bmcfee but have just realized that I had not been sent. Better late than never...]_
Sounds good to me. Most of what I need right now was already fixed in #307 anyway.
Then, what do you think about @dpwe 's experiments on moving the impulse location and getting a result that is not covariant with translation ?
